src/common_subset.rs

```diff
@@ -116,57 +117,117 @@ impl<NodeUid: Clone + Debug + Display + Eq + Hash + Ord> CommonSubset<NodeUid> {
-                                queue.into_iter().map(Output::Broadcast).collect()
-                            })
-                            .map_err(Error::from)
-                    } else {
-                        Err(Error::NoSuchBroadcastInstance)
-                    }
-                };
-                if instance_result.is_some() {
-                    self.on_broadcast_result(&uid)?;
-                }
-                input_result
+        input_message: Input<NodeUid>,
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
+        match input_message {
+            Input::Broadcast(uid, bmessage) => self.on_input_broadcast(&uid, bmessage),
+
+            Input::Agreement {
+                message_sender_id,
+                element_proposer_id,
+                message,
+            } => self.on_input_agreement(message_sender_id, element_proposer_id, &message),
+        }
+    }
+
+    fn on_input_broadcast(
+        &mut self,
+        uid: &NodeUid,
+        bmessage: BroadcastMessage<ProposedValue>,
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
+        let mut instance_result = None;
+        let input_result: Result<VecDeque<Output<NodeUid>>, Error> = {
+            if let Some(broadcast_instance) = self.broadcast_instances.get(&uid) {
+                broadcast_instance
+                    .handle_broadcast_message(&uid, bmessage)
+                    .map(|(opt_value, queue)| {
+                        instance_result = opt_value;
+                        queue.into_iter().map(Output::Broadcast).collect()
+                    })
+                    .map_err(Error::from)
+            } else {
+                Err(Error::NoSuchBroadcastInstance)
             }
-            Input::Agreement(_uid, _message) => {
-                // FIXME: send the message to the Agreement instance and
-                // conditionally call `on_agreement_output`
+        };
+        let mut opt_message: Option<AgreementMessage> = None;
+        if let Some(value) = instance_result {
+            self.broadcast_results.insert(uid.clone(), value);
+            opt_message = self.on_broadcast_result(&uid)?;
+        }
+        input_result.map(|mut queue| {
+            if let Some(agreement_message) = opt_message {
+                // Append the message to agreement nodes to the common output queue.
+                queue.push_back(Output::Agreement(agreement_message))
+            }
+            (None, queue)
+        })
+    }
+
+    fn on_input_agreement(
+        &mut self,
+        msg_sender_id: NodeUid,
+        element_proposer_id: NodeUid,
+        amessage: &AgreementMessage,
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
+        // The result defaults to error.
+        let mut result = Err(Error::NoSuchAgreementInstance);
+
+        // Send the message to the local instance of Agreement
+        if let Some(mut agreement_instance) = self.agreement_instances.get_mut(&element_proposer_id)
```

Comment:
Instead of the mutable `result`, you could just return here on error:

```rust
let agreement_instance = self.agreement_instances
    .get_mut(&element_proposer_id)
    .ok_or(Error::NoSuchAgreementInstance)?;
```

Reply:
Actually the mutable result helps to avoid a multiple mutable borrow conflict.